### ISSUE TITLE
Read tuple and array

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal-db#10]
Once and if we merge https://github.com/crystal-lang/crystal-db/pull/9 we can think of ways to easily read a row as a tuple, and also read many rows as an array, and even support a mapping, though mappings are more complex because one would need to use the correct names in column for that to work.

What I have so far is this:

```cr
struct Tuple
  def self.new(rs : DB::ResultSet)
    {% begin %}
      Tuple.new(
        {% for type in T %}
          rs.read({{type}}),
        {% end %}
      )
    {% end %}
  end
end

struct NamedTuple
  def self.new(rs : DB::ResultSet)
    {% begin %}
      NamedTuple.new(
        {% for key, type in T %}
          {{key.stringify}}: rs.read({{type}}),
        {% end %}
      )
    {% end %}
  end
end

class Array
  def self.query(db : DB::Database, query : String, *args)
    ary = new
    db.query(query, *args) do |rs|
      rs.each do
        ary << T.new(rs)
      end
    end
    ary
  end
end
```

With that, one can write:

```cr
DB.open "sqlite3://%3Amemory%3A" do |db|
  db.exec "create table contacts (name string, age integer)"
  db.exec "insert into contacts values (?, ?)", "John Doe", 30
  db.exec "insert into contacts values (?, ?)", "Foo Bar", 40

  db.query "select name, age from contacts order by age desc" do |rs|
    rs.each do
      p Tuple(String, Int32).new(rs)
      # => {"Foo Bar", 40}
      # => {"John Doe", 30}
    end
  end

  db.query "select name, age from contacts order by age desc" do |rs|
    rs.each do
      p NamedTuple(name: String, age: Int32).new(rs)
      # => {name: "Foo Bar", age: 40}
      # => {name: "John Doe", age: 30}
    end
  end

  ary = Array({String, Int32}).query(db, "select name, age from contacts order by age desc")
  p ary # => [{"Foo Bar", 40}, {"John Doe", 30}]

  ary = Array({name: String, age: Int32}).query(db, "select name, age from contacts order by age desc")
  p ary # => [{name: "Foo Bar", age: 40}, {name: "John Doe", age: 30}]
end
```
 
I'm still not sure it reads well, maybe we can add methods to `DB::Database` and `DB::ResultSet` instead, passing the types that we want. On the other hand, this is similar to JSON and YAML, where you put the type first:

```cr
Array(Int32).from_json(string)
# instead of:
# JSON.parse(string, Array(Int32))
```

Thoughts?

**Comment by bcardiff** [comment]
👍 for tuple
😕 for named tuple, since name can have space I would expect those to read by the name of the next column. consume the result set in order but fill the struct by name.

I kind of prefer `#from` rather than `#new`

I have bigger concerns regarding the name and location of `Array#query` you propose.

For mapping we have #2 standing by.

**Comment by asterite** [comment]
Closed in favor of #9
